File: src/system/refresher.ts
/**
 * src/system/refresher.ts — Refresher & Cache Invalidation Engine
 */

import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { execSync } from 'node:child_process';
import type { RefreshOptions, RefreshReport } from './types.ts';

export class Refresher {
  private projectDir: string;
  private homeDir: string;

  constructor(projectDir: string = '.') {
    this.projectDir = path.resolve(projectDir);
    this.homeDir = os.homedir();
  }

  private removeDirRecursive(dirPath: string): { count: number; bytes: number } {
    let count = 0;
    let bytes = 0;
    if (!fs.existsSync(dirPath)) return { count, bytes };

    try {
      const entries = fs.readdirSync(dirPath, { withFileTypes: true });
      for (const entry of entries) {
        const full = path.join(dirPath, entry.name);
        if (entry.isDirectory()) {
          const res = this.removeDirRecursive(full);
          count += res.count;
          bytes += res.bytes;
        } else {
          try {
            const stat = fs.statSync(full);
            bytes += stat.size;
            fs.unlinkSync(full);
            count++;
          } catch {
            // Ignore
          }
        }
      }
      fs.rmdirSync(dirPath);
    } catch {
      // Ignore
    }
    return { count, bytes };
  }

  private clearBytecodeCaches(): { items: string[]; bytes: number } {
    const items: string[] = [];
    let bytes = 0;

    const findPycache = (dir: string) => {
      if (!fs.existsSync(dir)) return;
      try {
        const entries = fs.readdirSync(dir, { withFileTypes: true });
        for (const entry of entries) {
          const full = path.join(dir, entry.name);
          if (entry.isDirectory()) {
            if (entry.name === '__pycache__' || entry.name === '.pytest_cache') {
              const res = this.removeDirRecursive(full);
              items.push(path.relative(this.projectDir, full));
              bytes += res.bytes;
            } else if (!entry.name.startsWith('.') && entry.name !== 'node_modules') {
              findPycache(full);
            }
          }
        }
      } catch {
        // Ignore
      }
    };

    findPycache(this.projectDir);
    return { items, bytes };
  }

  private cleanStaleLocks(): string[] {
    const cleared: string[] = [];
    const lockFiles = [
      path.join(this.projectDir, '.lock'),
      path.join(this.projectDir, '.git', 'index.lock'),
      path.join(this.projectDir, '.git', 'refs', 'heads', 'main.lock')
    ];

    for (const lf of lockFiles) {
      if (fs.existsSync(lf)) {
        try {
          const stat = fs.statSync(lf);
          // If older than 5 minutes, consider stale
          if (Date.now() - stat.mtimeMs > 5 * 60 * 1000) {
            fs.unlinkSync(lf);
            cleared.push(path.relative(this.projectDir, lf));
          }
        } catch {
          // Ignore
        }
      }
    }
    return cleared;
  }

  public refresh(options: RefreshOptions = {}): RefreshReport {
    const startTime = Date.now();
    const clearedItems: string[] = [];
    let freedBytes = 0;
    const messages: string[] = [];
    const hostsSynced: string[] = [];

    // 1. Clear Bytecode
    if (options.clearBytecode !== false) {
      const { items, bytes } = this.clearBytecodeCaches();
      clearedItems.push(...items);
      freedBytes += bytes;
      if (items.length > 0) {
        messages.push(`✓ Cleared ${items.length} Python bytecode cache directory(ies)`);
      }
    }

    // 2. Clean Stale Lockfiles
    if (options.cleanLockfiles !== false) {
      const locks = this.cleanStaleLocks();
      clearedItems.push(...locks);
      if (locks.length > 0) {
        messages.push(`✓ Removed ${locks.length} stale lockfile(s)`);
      }
    }

    // 3. Reset Risk Scores
    if (options.resetRiskScores) {
      const riskPath = path.join(this.projectDir, 'risk-scores.json');
      if (fs.existsSync(riskPath)) {
        try {
          fs.unlinkSync(riskPath);
          clearedItems.push('risk-scores.json');
          messages.push(`✓ Reset predictive debugging risk scores cache`);
        } catch {
          // Ignore
        }
      }
    }

    // 4. Rebuild Skills Index
    let skillsReindexed = false;
    if (options.rebuildSkillsIndex !== false) {
      try {
        execSync('python3 core/skills_indexer.py index', { cwd: this.projectDir, stdio: 'pipe' });
        skillsReindexed = true;
        messages.push(`✓ Rebuilt skills-index.json and skills-index.toon`);
      } catch (err: any) {
        messages.push(`! Skills re-indexing notice: ${err.message}`);
      }
    }

    // 5. Sync Integrations
    let integrationsSynced = false;
    if (options.syncIntegrations !== false) {
      try {
        const script = `from core.integrations import IntegrationInstaller; IntegrationInstaller('${this.projectDir}').provision_all()`;
        execSync(`python3 -c "${script}"`, { cwd: this.projectDir, stdio: 'pipe' });
        integrationsSynced = true;
        messages.push(`✓ Synchronized supportive integrations (Ponytail, TOON, Fable, Caveman)`);
      } catch (err: any) {
        messages.push(`! Integrations sync notice: ${err.message}`);
      }
    }

    // 6. Sync Host Skills
    if (options.syncHostSkills) {
      const hostDirs = [
        path.join(this.homeDir, '.claude', 'skills', 'agentic-workflow'),
        path.join(this.homeDir, '.gemini', 'config', 'skills', 'agentic-workflow'),
        path.join(this.homeDir, '.cursor', 'skills', 'agentic-workflow'),
        path.join(this.homeDir, '.agents', 'skills', 'agentic-workflow')
      ];

      for (const hDir of hostDirs) {
        if (fs.existsSync(hDir)) {
          try {
            // Update SKILL.md and marketplace.json
            fs.copyFileSync(path.join(this.projectDir, 'SKILL.md'), path.join(hDir, 'SKILL.md'));
            hostsSynced.push(hDir);
          } catch {
            // Ignore
          }
        }
      }
      if (hostsSynced.length > 0) {
        messages.push(`✓ Updated skill definitions across ${hostsSynced.length} host environments`);
      }
    }

    const durationMs = Date.now() - startTime;
    return {
      success: true,
      clearedItems,
      freedBytes,
      skillsReindexed,
      integrationsSynced,
      hostsSynced,
      durationMs,
      messages
    };
  }
}

File: core/system/refresher.py
"""
core/system/refresher.py — Python Refresher & Cache Invalidation Engine
"""

import os
import shutil
import time
import subprocess
from pathlib import Path
from typing import Dict, Any, List


class Refresher:
    def __init__(self, project_dir: str = "."):
        self.project_dir = Path(project_dir).resolve()
        self.home_dir = Path.home()

    def clear_bytecode(self) -> Dict[str, Any]:
        cleared = []
        freed = 0

        for p in self.project_dir.rglob("__pycache__"):
            if "node_modules" in str(p) or ".git" in str(p):
                continue
            try:
                for f in p.rglob("*"):
                    if f.is_file():
                        freed += f.stat().st_size
                shutil.rmtree(p, ignore_errors=True)
                cleared.append(str(p.relative_to(self.project_dir)))
            except Exception:
                _cleared = False

        for p in self.project_dir.rglob(".pytest_cache"):
            if "node_modules" in str(p):
                continue
            try:
                shutil.rmtree(p, ignore_errors=True)
                cleared.append(str(p.relative_to(self.project_dir)))
            except Exception:
                _cleared = False

        return {"cleared": cleared, "freed_bytes": freed}

    def clean_stale_locks(self) -> List[str]:
        cleared = []
        now = time.time()
        for lock_name in [".lock", ".git/index.lock"]:
            lf = self.project_dir / lock_name
            if lf.exists():
                try:
                    if now - lf.stat().st_mtime > 300:
                        lf.unlink()
                        cleared.append(lock_name)
                except Exception:
                    _unlinked = False
        return cleared

    def refresh(self, clear_bytecode: bool = True, rebuild_index: bool = True, sync_integrations: bool = True) -> Dict[str, Any]:
        start = time.time()
        messages = []
        cleared_items = []
        freed_bytes = 0

        if clear_bytecode:
            bc = self.clear_bytecode()
            cleared_items.extend(bc["cleared"])
            freed_bytes += bc["freed_bytes"]
            if bc["cleared"]:
                messages.append(f"✓ Cleared {len(bc['cleared'])} bytecode directory(ies)")

        locks = self.clean_stale_locks()
        cleared_items.extend(locks)
        if locks:
            messages.append(f"✓ Removed {len(locks)} stale lockfile(s)")

        skills_reindexed = False
        if rebuild_index:
            try:
                subprocess.run(
                    "python3 core/skills_indexer.py index",
                    shell=True,
                    cwd=self.project_dir,
                    capture_output=True
                )
                skills_reindexed = True
                messages.append("✓ Rebuilt skills mesh index")
            except Exception as e:
                messages.append(f"! Skills re-indexing notice: {e}")

        integrations_synced = False
        if sync_integrations:
            try:
                from core.integrations import IntegrationInstaller
                IntegrationInstaller(str(self.project_dir)).provision_all()
                integrations_synced = True
                messages.append("✓ Synchronized supportive integrations")
            except Exception as e:
                messages.append(f"! Integrations sync notice: {e}")

        duration_ms = int((time.time() - start) * 1000)
        return {
            "success": True,
            "cleared_items": cleared_items,
            "freed_bytes": freed_bytes,
            "skills_reindexed": skills_reindexed,
            "integrations_synced": integrations_synced,
            "duration_ms": duration_ms,
            "messages": messages
        }
